### PR TITLE
NVMe Direct: Added 5 testcases for Azure and Hyper-V

### PR DESCRIPTION
**Pull request by adriansuhov**
The test cases added:
- Basic validation
- Functional
- Functional (Expected fails)
- Performance FIO Single Disk
- Performance FIO Multi Disk (RAID 0)

Also added 2 ps1 scripts for Hyper-V: one for attaching NVME devices and the other to detach them. All the details of the test scripts are in the description of each test scripts.

## Hyper-V run
[LISAv2 Test Results Summary]
Test Run On           : 12/18/2018 17:49:15
VHD Under Test     : D:\Users\v-adsuho\Documents\ubuntu_16.04.5-Copy.vhdx
Total Test Cases      : 5 (5 Pass, 0 Fail, 0 Abort)
Total Time (dd:hh:mm) : 0:0:36
XML File              : TestConfiguration

   ID TestCaseName                                                 TestResult TestDuration(in minutes)
------------------------------------------------------------------------------------------------------
    1 NVME-BASIC-VALIDATION                                              PASS                 4.56
    2 NVME-FUNCTIONAL                                                    PASS                 4.75
    3 NVME-FUNCTIONAL-EXPECTED-FAILS                                     PASS                 2.87
    4 PERF-NVME-4K-IO-SINGLE                                             PASS                10.53
    5 PERF-NVME-4K-IO-MULTI                                              PASS                13.64

All the scripts have also been tested in Azure VMs, but not with lisav2. This is because VMs with nvme disks are yet to be available.

Also retested regular FIO test to make sure that nothing was broken on the regular storage performance test.

   ID TestCaseName                                                 TestResult TestDuration(in minutes)
------------------------------------------------------------------------------------------------------
    1 PERF-STORAGE-4K-IO                                                 PASS                14.98

### CODE REVIEW

File: Testscripts/Linux/nvme_basic_validation.sh [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache License.
+#
+# Description:
+# Basic NVME test script: 
+# - It checks if nvme namespace is available in /dev (e.g. /dev/nvme0n1)
+# - Compares the number of namespaces with the number of devices 
+# - Installs nvme-cli and checks the listing of the device 
+# - Azure only: It compares availalbe namespaces with the number of vCPUs
```
Comment (chvalean) [inline Testscripts/Linux/nvme_basic_validation.sh]: typo in availalbe

File: Testscripts/Linux/nvme_functional_expected_fails.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    # The list 'expected_fail_cmd_list' can be expanded if required
+    for nvme_cmd in "${expected_fail_cmd_list[@]}"; do
+        nvme $nvme_cmd /dev/"$namespace"
+        if [ $? -eq 0 ]; then
+            LogErr "The command 'nvme ${nvme_cmd} $namespace' should have failed!"
+            SetTestStateFailed
+            exit 0
+        fi
+    done
+    UpdateSummary "All the comands on ${namespace} failed as expected!"
```
Comment (chvalean) [inline Testscripts/Linux/nvme_functional_expected_fails.sh]: commands*

File: Testscripts/Windows/SETUP-Add-NVME-Disk.ps1 [hunk 3]
```diff
@@ -0,0 +1,114 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache License.
+
+<#
+.Description
+    This setup script adds a specified number of NVME disks to a test VM
+    It will first search if there are any VMs using NVME disks and detach
+    the disk(s). After all the NVME disks are offlline and unnasigned to
```
Comment (chvalean) [inline Testscripts/Windows/SETUP-Add-NVME-Disk.ps1]: offline*

**Comment by adriansuhov** [comment]
The comments have been addressed. 
I also have Azure results:

[LISAv2 Test Results Summary]
Test Run On           : 12/19/2018 08:30:33
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 5 (4 Pass, 0 Fail, 1 Abort)
Total Time (dd:hh:mm) : 0:0:39
XML File              : TestConfiguration

   ID TestCaseName                                                 TestResult TestDuration(in minutes)
------------------------------------------------------------------------------------------------------
    1 NVME-BASIC-VALIDATION                                              PASS                 6.52
    2 NVME-FUNCTIONAL                                                    PASS                 6.71
    3 NVME-FUNCTIONAL-EXPECTED-FAILS                                     PASS                 5.93
    4 PERF-NVME-4K-IO-SINGLE                                             PASS                16.63
    5 PERF-NVME-4K-IO-MULTI                                           ABORTED                 0.41

**PERF-NVME-4K-IO-MULTI failed because of the lack of resources available on Azure. Here is a rerun after resources were avaialable:**

ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 1 (1 Pass, 0 Fail, 0 Abort)
Total Time (dd:hh:mm) : 0:0:24
XML File              : TestConfiguration

   ID TestCaseName                                                 TestResult TestDuration(in minutes)
------------------------------------------------------------------------------------------------------
    1 PERF-NVME-4K-IO-MULTI                                              PASS                19.24

File: Testscripts/Linux/nvme_basic_validation.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache License.
+#
+# Description:
+# Basic NVME test script: 
+# - It checks if nvme namespace is available in /dev (e.g. /dev/nvme0n1)
+# - Compares the number of namespaces with the number of devices
```
Comment (chvalean) [inline Testscripts/Linux/nvme_basic_validation.sh]: remove blank spaces at end of lines

File: Testscripts/Linux/nvme_basic_validation.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        LogErr "nvme namespace ${namespace} was not detected by nvme-cli"
+        SetTestStateFailed
+        exit 0
+    else
+        echo "nvme namespace ${namespace} was detected by nvme-cli"
+        UpdateSummary "${namespace} detected!"
+    fi
+done
+
+# Only for Azure. Check if namespace count is in line with expected number of vCPUs.
```
Comment (chvalean) [inline Testscripts/Linux/nvme_basic_validation.sh]: white space at EOL

File: Testscripts/Linux/nvme_basic_validation.sh [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        UpdateSummary "${namespace} detected!"
+    fi
+done
+
+# Only for Azure. Check if namespace count is in line with expected number of vCPUs. 
+# The expected ratio is 1 nvme disk to 8 vCPU
+if [ ! -e platform.txt ]; then
+    expected_namespace_count=$(($(nproc) / 8))
+    if [ "$namespace_count" -ne "$expected_namespace_count" ]; then
+        LogErr "CPU to nvme devices is not 8/1!"
```
Comment (chvalean) [inline Testscripts/Linux/nvme_basic_validation.sh]: CPU to NVMe devices ratio*

File: XML/TestCases/FunctionalTests-NVME.xml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        <OverrideVMSize>Standard_L16s_v2</OverrideVMSize>
+        <testParameters>
+            <param>HYPERV_DISK_COUNT=2</param>
+        </testParameters>
+        <cleanupScript>.\Testscripts\Windows\SETUP-Remove-NVME-Disk.ps1</cleanupScript>
+        <Platform>Azure,HyperV</Platform>
+        <Category>Functional</Category>
+        <Area>NVME</Area>
+        <Tags>nvme</Tags>
+        <Priority>1</Priority>
```
Comment (chvalean) [inline XML/TestCases/FunctionalTests-NVME.xml]: P0

File: XML/TestCases/FunctionalTests-NVME.xml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    </test>
+    <test>
+        <testName>NVME-FUNCTIONAL</testName>
```
Comment (chvalean) [inline XML/TestCases/FunctionalTests-NVME.xml]: must change to something more descriptive

File: XML/TestCases/FunctionalTests-NVME.xml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        <OverrideVMSize>Standard_L8s_v2</OverrideVMSize>
+        <testParameters>
+            <param>HYPERV_DISK_COUNT=1</param>
+        </testParameters>
+        <cleanupScript>.\Testscripts\Windows\SETUP-Remove-NVME-Disk.ps1</cleanupScript>
+        <Platform>Azure,HyperV</Platform>
+        <Category>Functional</Category>
+        <Area>NVME</Area>
+        <Tags>nvme</Tags>
+        <Priority>1</Priority>
```
Comment (chvalean) [inline XML/TestCases/FunctionalTests-NVME.xml]: P0

File: XML/TestCases/FunctionalTests-NVME.xml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
```
Comment (chvalean) [inline XML/TestCases/FunctionalTests-NVME.xml]: P0

**Comment by adriansuhov** [comment]
Changes made to address comments

**Pull request merged**